domain/storage/state/state.go

```diff
@@ -0,0 +1,332 @@
+		err := tx.Query(ctx, selectUUIDStmt, dbPool).Get(&dbPool)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(domain.CoerceError(err))
+		}
+		if err == nil {
+			return fmt.Errorf("storage pool %q %w", pool.Name, errors.AlreadyExists)
+		}
+		poolUUID := uuid.MustNewUUID().String()
+
+		if err := upsertStoragePool(ctx, tx, poolUUID, pool.Name, pool.Provider); err != nil {
+			return errors.Annotate(domain.CoerceError(err), "creating storage pool")
+		}
+
+		if err := updatePoolAttributes(ctx, tx, poolUUID, pool.Attrs); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "creating storage pool %s attributes", poolUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func upsertStoragePool(ctx context.Context, tx *sqlair.TX, poolUUID string, name string, providerType string) error {
+	if name == "" {
+		return fmt.Errorf("storage pool name cannot be empty%w", errors.Hide(errors.NotValid))
+	}
+	if providerType == "" {
+		return fmt.Errorf("storage pool type cannot be empty%w", errors.Hide(errors.NotValid))
+	}
+
+	dbPool := StoragePool{
+		ID:           poolUUID,
+		Name:         name,
+		ProviderType: providerType,
+	}
+
+	insertQuery := `
+INSERT INTO storage_pool (uuid, name, type)
+VALUES (
+    $StoragePool.uuid,
+    $StoragePool.name,
+    $StoragePool.type
+)
+ON CONFLICT(uuid) DO UPDATE SET name=excluded.name,
+                                type=excluded.type
+`
+
+	insertStmt, err := sqlair.Prepare(insertQuery, StoragePool{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = tx.Query(ctx, insertStmt, dbPool).Run()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func updatePoolAttributes(ctx context.Context, tx *sqlair.TX, storagePoolUUID string, attr domainstorage.Attrs) error {
+	// Delete any keys no longer in the attributes map.
+	// TODO(wallyworld) - use sqlair NOT IN operation
+	deleteQuery := fmt.Sprintf(`
+DELETE FROM  storage_pool_attribute
+WHERE        storage_pool_uuid = $M.uuid
+-- AND          key NOT IN (?)
+`)
+
+	deleteStmt, err := sqlair.Prepare(deleteQuery, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := tx.Query(ctx, deleteStmt, sqlair.M{"uuid": storagePoolUUID}).Run(); err != nil {
+		return errors.Trace(domain.CoerceError(err))
+	}
+
+	insertQuery := `
+INSERT INTO storage_pool_attribute
+VALUES (
+    $poolAttribute.storage_pool_uuid,
+    $poolAttribute.key,
+    $poolAttribute.value
+)
+ON CONFLICT(storage_pool_uuid, key) DO UPDATE SET key=excluded.key,
+                                                      value=excluded.value
+`
+	insertStmt, err := sqlair.Prepare(insertQuery, poolAttribute{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for key, value := range attr {
+		if err := tx.Query(ctx, insertStmt, poolAttribute{
+			ID:    storagePoolUUID,
+			Key:   key,
+			Value: value,
+		}).Run(); err != nil {
+			return errors.Trace(domain.CoerceError(err))
```

Comment:
I think it should be the callers responsibility to Coerce as we are still in the state layer. More two cents.